{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import distance\n",
    "import matplotlib.pyplot as plt\n",
    "from itertools import combinations,combinations_with_replacement\n",
    "import time\n",
    "import os"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Utility functions for parsing FASTQs\n",
    "def get_ID_from_line(line):\n",
    "    return line.strip().split()[0]\n",
    "\n",
    "def get_seq_from_line(line):\n",
    "    return line.strip()\n",
    "\n",
    "def get_sequence_generator(fname):\n",
    "    with open(fname,'r') as f:\n",
    "        while True:\n",
    "            for i in range(4):\n",
    "                try:\n",
    "                    line = next(f)\n",
    "                except StopIteration:\n",
    "                    return\n",
    "                if i == 0: ID = get_ID_from_line(line)\n",
    "                if i == 1: seq = get_seq_from_line(line)\n",
    "                if i == 3: yield (ID, seq)\n",
    "\n",
    "# Creates dictionary of sequences a max Hamming distance away from input sequence, \n",
    "# allowing clustering of similar barcodes\n",
    "def get_mutation_dict(name,seq,hamming_distance = 1):\n",
    "    \"\"\" makes a dictionary that maps sequence variants to the feature,hamming distance \"\"\"\n",
    "    mutation_dict = {seq: (name,0)} # set up empty dictionary\n",
    "\n",
    "    def mutate_seq(seq,dist):\n",
    "        \"\"\" adds mutated sequences of set hamming distance to list\"\"\"\n",
    "        position_sets = list(combinations(range(len(seq)),dist))\n",
    "        base_sets = list(combinations_with_replacement('AGCTN',dist))\n",
    "        for positions in position_sets:\n",
    "            for bases in base_sets:\n",
    "                mutated_seq = seq # create copy of sequence\n",
    "                #\n",
    "                for pos,b in zip(positions,bases):\n",
    "                    mutated_seq = mutated_seq[:pos] + b + mutated_seq[pos + 1:]\n",
    "                # if there isn't a better sequence already stored\n",
    "                if not mutated_seq in mutation_dict:\n",
    "                    mutation_dict[mutated_seq] = (name,dist)\n",
    "\n",
    "    for dist in range(1,hamming_distance + 1):\n",
    "        mutate_seq(seq,dist)\n",
    "\n",
    "    return mutation_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'\\n    max_index = len(seq)-len(pre)-bc_length-len(post)\\n    for i in range(max_index+1):\\n        current = seq[i:i+len(pre)]\\n        if current in pre_hash.keys():\\n            bc_start = i+len(pre)\\n            post_check = seq[bc_start+bc_length:bc_start+bc_length+len(post)]\\n            if post_check in post_hash.keys():\\n                bc = seq[bc_start:bc_start+bc_length]\\n                return [bc, current, post_check, seq]\\n\\n\\n    return None\\n'"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Returns BC; if barcode is not found, returns None\n",
    "def find_barcode(seq, pre_hash, post_hash, bc_length, max_dist):\n",
    "    \n",
    "    # Constraining positions of flanking regions and BCs:\n",
    "    pre = list(pre_hash.keys())[0]\n",
    "    post = list(post_hash.keys())[0]\n",
    "    BC_index = 37 # updated\n",
    "    pre_index = 17\n",
    "    post_index = 75\n",
    "    \n",
    "    pre_check = seq[pre_index:BC_index]\n",
    "    post_check = seq[BC_index+bc_length:post_index]\n",
    "    \n",
    "    if pre_check in pre_hash and post_check in post_hash:\n",
    "        return seq[BC_index:BC_index+bc_length]\n",
    "    else:\n",
    "        return None\n",
    "\n",
    "'''\n",
    "    max_index = len(seq)-len(pre)-bc_length-len(post)\n",
    "    for i in range(max_index+1):\n",
    "        current = seq[i:i+len(pre)]\n",
    "        if current in pre_hash.keys():\n",
    "            bc_start = i+len(pre)\n",
    "            post_check = seq[bc_start+bc_length:bc_start+bc_length+len(post)]\n",
    "            if post_check in post_hash.keys():\n",
    "                bc = seq[bc_start:bc_start+bc_length]\n",
    "                return [bc, current, post_check, seq]\n",
    "\n",
    "\n",
    "    return None\n",
    "'''"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Returns a dictionary of {barcodes : counts}\n",
    "# Inputs = input_file name, 5' flanking region, 3' flanking region, bc_length, hamming distance for flanking regions\n",
    "def count_barcodes(input_file, bc_5p, bc_3p, bc_length, max_dist):\n",
    "\n",
    "    bc_5p_hash = get_mutation_dict(bc_5p, bc_5p, max_dist)\n",
    "    bc_3p_hash = get_mutation_dict(bc_3p, bc_3p, max_dist)\n",
    "\n",
    "    bc_freq = {}\n",
    "    seq_count = 0\n",
    "    flank_count = 0\n",
    "    t0 = time.process_time()\n",
    "\n",
    "    for ID,seq in get_sequence_generator(input_file):\n",
    "        seq_count += 1\n",
    "\n",
    "        if seq_count % 1000000 == 0:\n",
    "            print(str(seq_count) + ' sequences processed. (Elapsed time: ' + str(time.process_time()-t0) + ' secs.)')\n",
    "            \n",
    "        bc = find_barcode(seq, bc_5p_hash, bc_3p_hash, bc_length, max_dist)\n",
    "        if bc != None:\n",
    "            flank_count += 1\n",
    "            if bc in bc_freq:\n",
    "                bc_freq[bc] += 1\n",
    "            else:\n",
    "                bc_freq[bc] = 1\n",
    "            '''\n",
    "                collapsed = False\n",
    "                for k in bc_freq:\n",
    "                    if distance.hamming(k,bc) < dist_threshold:\n",
    "                        bc = k\n",
    "                        collapsed = True\n",
    "                        break\n",
    "                if collapsed:\n",
    "                    bc_freq[bc] += 1\n",
    "                else:\n",
    "                    bc_freq[bc] = 1\n",
    "            '''\n",
    "    \n",
    "    bc_purity = (flank_count / seq_count) * 100.0\n",
    "    \n",
    "    print(str(seq_count) + ' total sequences processed.')\n",
    "    print('Flanking regions found in ' + str(round(bc_purity,2)) + '% of sequences.')\n",
    "    \n",
    "    df = pd.Series(bc_freq)\n",
    "    return df.sort_values(ascending = False)\n",
    "    #df = pd.DataFrame.from_dict(bc_freq, orient='index', columns = ['Counts'])\n",
    "    #return df.sort_values(by = 'Counts', ascending = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Collapse similar barcodes post-hoc\n",
    "# Input = barcode dataframe/series containing sorted raw counts of uncollapsed barcodes, and Hamming dist threshold\n",
    "# Returns sorted dataframe containing collapsed barcode counts and dictionary of {collapsed BC : [original BCs]}\n",
    "def collapse_barcodes(barcode_df, max_dist):\n",
    "    barcodes = list(barcode_df.index) # retains sorted order\n",
    "    bc_dict = barcode_df.to_dict()\n",
    "    i = 0\n",
    "    collapsed_total = []\n",
    "    collapsed_dict = {}\n",
    "    while i + len(collapsed_total) < len(barcodes):\n",
    "        bc = barcodes[i]\n",
    "        if bc not in collapsed_total:\n",
    "            similar_hits = []\n",
    "            similar_bcs = list(get_mutation_dict(bc, bc, max_dist).keys())[1:]\n",
    "            for similar_bc in similar_bcs:\n",
    "                if similar_bc in bc_dict:\n",
    "                    similar_hits.append(similar_bc)\n",
    "                    similar_bc_count = bc_dict.pop(similar_bc)\n",
    "                    bc_dict[bc] += similar_bc_count\n",
    "                    collapsed_total.append(similar_bc)\n",
    "            if len(similar_hits) > 0:\n",
    "                collapsed_dict[bc] = similar_hits\n",
    "        i += 1\n",
    "    print(str(len(collapsed_total)) + ' barcodes collapsed.')\n",
    "    print(str(len(barcodes)) + ' starting barcodes folded into ' + str(len(bc_dict)) + '.')\n",
    "    df = pd.DataFrame.from_dict(bc_dict, orient='index', columns = ['Counts'])\n",
    "    collapsed_bc_df = pd.DataFrame.from_dict(collapsed_dict, orient='index')\n",
    "    return df.sort_values(by = 'Counts', ascending = False), collapsed_bc_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save output file\n",
    "def save_output(sample, df, suffix = None):\n",
    "    if suffix == None:\n",
    "        suffix = ''\n",
    "    else:\n",
    "        suffix = '_' + suffix\n",
    "    output_file = './output/' + sample + suffix + '.xlsx'\n",
    "    df.to_excel(output_file)\n",
    "    return output_file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Processing file: Replicate2_2nd CD69+PD-1- sorted.fastq\n",
      "367348 total sequences processed.\n",
      "Flanking regions found in 57.11% of sequences.\n",
      "931 barcodes collapsed.\n",
      "34965 starting barcodes folded into 34034.\n",
      "Collapsed barcode counts saved in ./output/Replicate2_2nd CD69+PD-1- sorted.fastq.xlsx.\n",
      "Processing file: Replicate2_EGFP+ sorted.fastq\n",
      "1000000 sequences processed. (Elapsed time: 9.406338000000005 secs.)\n",
      "1332372 total sequences processed.\n",
      "Flanking regions found in 86.38% of sequences.\n",
      "6420 barcodes collapsed.\n",
      "106411 starting barcodes folded into 99991.\n",
      "Collapsed barcode counts saved in ./output/Replicate2_EGFP+ sorted.fastq.xlsx.\n",
      "Processing file: EGFP+ sorted.fastq\n",
      "622816 total sequences processed.\n",
      "Flanking regions found in 96.24% of sequences.\n",
      "12725 barcodes collapsed.\n",
      "141230 starting barcodes folded into 128505.\n",
      "Collapsed barcode counts saved in ./output/EGFP+ sorted.fastq.xlsx.\n",
      "Processing file: 1st CD69+ sorted.fastq\n",
      "1000000 sequences processed. (Elapsed time: 11.354753999999957 secs.)\n",
      "1041395 total sequences processed.\n",
      "Flanking regions found in 96.73% of sequences.\n",
      "13371 barcodes collapsed.\n",
      "38892 starting barcodes folded into 25521.\n",
      "Collapsed barcode counts saved in ./output/1st CD69+ sorted.fastq.xlsx.\n",
      "Processing file: Replicate2_1st CD69+PD-1- sorted.fastq\n",
      "201497 total sequences processed.\n",
      "Flanking regions found in 58.5% of sequences.\n",
      "595 barcodes collapsed.\n",
      "25882 starting barcodes folded into 25287.\n",
      "Collapsed barcode counts saved in ./output/Replicate2_1st CD69+PD-1- sorted.fastq.xlsx.\n",
      "Processing file: 2nd CD69+ sorted.fastq\n",
      "1000000 sequences processed. (Elapsed time: 9.536053999999922 secs.)\n",
      "1361257 total sequences processed.\n",
      "Flanking regions found in 98.31% of sequences.\n",
      "7535 barcodes collapsed.\n",
      "17793 starting barcodes folded into 10258.\n",
      "Collapsed barcode counts saved in ./output/2nd CD69+ sorted.fastq.xlsx.\n",
      "Processing file: 3rd CD69+ sorted.fastq\n",
      "1000000 sequences processed. (Elapsed time: 10.026125999999977 secs.)\n",
      "1137818 total sequences processed.\n",
      "Flanking regions found in 98.5% of sequences.\n",
      "5270 barcodes collapsed.\n",
      "11989 starting barcodes folded into 6719.\n",
      "Collapsed barcode counts saved in ./output/3rd CD69+ sorted.fastq.xlsx.\n",
      "Processing file: Unselected.fastq\n",
      "1000000 sequences processed. (Elapsed time: 10.290556000000151 secs.)\n",
      "2000000 sequences processed. (Elapsed time: 21.66030599999999 secs.)\n",
      "2128037 total sequences processed.\n",
      "Flanking regions found in 96.91% of sequences.\n",
      "55849 barcodes collapsed.\n",
      "742568 starting barcodes folded into 686719.\n",
      "Collapsed barcode counts saved in ./output/Unselected.fastq.xlsx.\n",
      "Processing file: Replicate2_3rd CD69+ sorted.fastq\n",
      "82938 total sequences processed.\n",
      "Flanking regions found in 92.97% of sequences.\n",
      "452 barcodes collapsed.\n",
      "11447 starting barcodes folded into 10995.\n",
      "Collapsed barcode counts saved in ./output/Replicate2_3rd CD69+ sorted.fastq.xlsx.\n",
      "Processing file: Replicate2_1st CD69+ sorted.fastq\n",
      "116132 total sequences processed.\n",
      "Flanking regions found in 92.26% of sequences.\n",
      "601 barcodes collapsed.\n",
      "16494 starting barcodes folded into 15893.\n",
      "Collapsed barcode counts saved in ./output/Replicate2_1st CD69+ sorted.fastq.xlsx.\n",
      "Processing file: Replicate2_Non-select.fastq\n",
      "1000000 sequences processed. (Elapsed time: 10.61799000000019 secs.)\n",
      "1023830 total sequences processed.\n",
      "Flanking regions found in 96.04% of sequences.\n",
      "6526 barcodes collapsed.\n",
      "533447 starting barcodes folded into 526921.\n",
      "Collapsed barcode counts saved in ./output/Replicate2_Non-select.fastq.xlsx.\n",
      "Processing file: Replicate2_2nd CD69+ sorted.fastq\n",
      "246948 total sequences processed.\n",
      "Flanking regions found in 54.56% of sequences.\n",
      "700 barcodes collapsed.\n",
      "27134 starting barcodes folded into 26434.\n",
      "Collapsed barcode counts saved in ./output/Replicate2_2nd CD69+ sorted.fastq.xlsx.\n",
      "Processing file: 1st CD69+PD-1- sorted.fastq\n",
      "1000000 sequences processed. (Elapsed time: 9.688626000000113 secs.)\n",
      "1189479 total sequences processed.\n",
      "Flanking regions found in 98.16% of sequences.\n",
      "5238 barcodes collapsed.\n",
      "10746 starting barcodes folded into 5508.\n",
      "Collapsed barcode counts saved in ./output/1st CD69+PD-1- sorted.fastq.xlsx.\n",
      "Processing file: Replicate2_PlasmidLib.fastq\n",
      "695357 total sequences processed.\n",
      "Flanking regions found in 96.11% of sequences.\n",
      "3131 barcodes collapsed.\n",
      "436573 starting barcodes folded into 433442.\n",
      "Collapsed barcode counts saved in ./output/Replicate2_PlasmidLib.fastq.xlsx.\n",
      "Processing file: 2nd CD69+PD-1- sorted.fastq\n",
      "1000000 sequences processed. (Elapsed time: 9.727789999999914 secs.)\n",
      "1315463 total sequences processed.\n",
      "Flanking regions found in 98.72% of sequences.\n",
      "2636 barcodes collapsed.\n",
      "5725 starting barcodes folded into 3089.\n",
      "Collapsed barcode counts saved in ./output/2nd CD69+PD-1- sorted.fastq.xlsx.\n"
     ]
    }
   ],
   "source": [
    "### MAIN ###\n",
    "\n",
    "#20 bp flanking regions, max Hamming = 1 for flanking regions and barcode collapsing\n",
    "bc_5p = 'TTGGAGAACCACCTTGTTGG'\n",
    "bc_3p = 'GTTTAAGAGCTAAGCTGGAA' # updated with Taeyoon's BC flanking sequences\n",
    "dist_threshold = 1\n",
    "bc_length = 18\n",
    "\n",
    "cwd = os.getcwd()\n",
    "for filename in os.listdir(cwd):\n",
    "    if filename.endswith('.fastq'):\n",
    "        print('Processing file: ' + filename)\n",
    "        uncollapsed_bcs = count_barcodes(filename, bc_5p, bc_3p, bc_length, dist_threshold)\n",
    "        collapsed_bc_counts, collapsed_bcs = collapse_barcodes(uncollapsed_bcs, dist_threshold)\n",
    "        output_file = save_output(filename, collapsed_bc_counts)\n",
    "        print('Collapsed barcode counts saved in ' + output_file + '.')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Checking collapsing results\n",
    "check = True\n",
    "for index, row in collapsed_bcs.iterrows():\n",
    "    for i in range(len(row)):\n",
    "        if row[i] != None:\n",
    "            check = distance.hamming(index,row[i]) <= 1\n",
    "print(check)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Uncollapsed analysis\n",
    "### MAIN ###\n",
    "\n",
    "#20 bp flanking regions, max Hamming = 1 for flanking regions and barcode collapsing\n",
    "bc_5p = 'TTGGAGAACCACCTTGTTGG'\n",
    "bc_3p = 'TTTGTACCCCGTATTCCGTT'\n",
    "dist_threshold = 1\n",
    "bc_length = 18\n",
    "\n",
    "cwd = os.getcwd()\n",
    "for filename in os.listdir(cwd):\n",
    "    if filename.endswith('.fastq'):\n",
    "        print('Processing file: ' + filename)\n",
    "        uncollapsed_bcs = count_barcodes(filename, bc_5p, bc_3p, bc_length, dist_threshold)\n",
    "        collapsed_bcs = collapse_barcodes(uncollapsed_bcs, dist_threshold)\n",
    "        output_file = save_output(filename, uncollapsed_bcs)\n",
    "        print('Collapsed barcode counts saved in ' + output_file + '.')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Returns table of barcode counts, frequencies, and BC ranks for input sample\n",
    "# By default processes collapsed reads, without a BC rank cutoff\n",
    "def read_barcode_table(sample, collapsed = True, rank_cutoff = None):\n",
    "    if collapsed:\n",
    "        sample += '_collapsed'\n",
    "    input_file = './output/' + sample + '.xlsx'\n",
    "    df = pd.read_excel(input_file, index_col = 0, header = 0, usecols = 'A:D', nrows = rank_cutoff)\n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Comparison of selected to control barcodes\n",
    "# Input: Sample names (as strings) of selected and control samples, Hamming distance cutoff to define overlapping\n",
    "# barcodes, and number of top barcodes in selected sample to use as a query (default = None, processes all BCs)\n",
    "# Returns dataframe of: shared BCs, read counts, frequencies, and BC ranks in both selected and control datasets\n",
    "def barcode_overlap(selected_sample, control_sample, max_dist, read_cutoff, rank_cutoff = None, collapsed = True):\n",
    "    selected_df = read_barcode_table(selected_sample, collapsed = collapsed, rank_cutoff = rank_cutoff)\n",
    "    control_df = read_barcode_table(control_sample, collapsed = collapsed)\n",
    "    selected_bcs = selected_df.to_dict(orient='index')\n",
    "    control_bcs = control_df.to_dict(orient='index')\n",
    "    shared_bcs = []\n",
    "    for bc in selected_bcs:\n",
    "        similar_bcs = list(get_mutation_dict(bc, bc, max_dist).keys())\n",
    "        for bc_check in similar_bcs:\n",
    "            if bc_check in control_bcs:\n",
    "                selected_count = selected_df['Counts'][bc]\n",
    "                selected_freq = selected_df['Freq'][bc]\n",
    "                selected_rank = selected_df['BC Num'][bc]\n",
    "                control_count = control_df['Counts'][bc_check]\n",
    "                control_freq = control_df['Freq'][bc_check]\n",
    "                control_rank = control_df['BC Num'][bc_check]\n",
    "                if selected_count >= read_cutoff and control_count >= read_cutoff:\n",
    "                    shared_bcs.append([bc, selected_count, selected_freq, selected_rank, \n",
    "                                       bc_check, control_count, control_freq, control_rank])\n",
    "    output = pd.DataFrame(data = shared_bcs, columns = ['Selected BC', 'Selected Counts', 'Selected Freq', 'Selected Rank',\n",
    "                                                       'Control BC', 'Control Counts', 'Control Freq', 'Control Rank'])\n",
    "    return output.sort_values(by='Selected Rank')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Selected BC</th>\n",
       "      <th>Selected Counts</th>\n",
       "      <th>Selected Freq</th>\n",
       "      <th>Selected Rank</th>\n",
       "      <th>Control BC</th>\n",
       "      <th>Control Counts</th>\n",
       "      <th>Control Freq</th>\n",
       "      <th>Control Rank</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AAAAGCGCCGACAAACCC</td>\n",
       "      <td>2271</td>\n",
       "      <td>0.000698</td>\n",
       "      <td>2</td>\n",
       "      <td>AAAAGCGCCGACAAACCC</td>\n",
       "      <td>173</td>\n",
       "      <td>0.000046</td>\n",
       "      <td>8329</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GAGGACGCCACAAACAGA</td>\n",
       "      <td>1524</td>\n",
       "      <td>0.000468</td>\n",
       "      <td>7</td>\n",
       "      <td>GAGGACGCCACAAACAGA</td>\n",
       "      <td>83</td>\n",
       "      <td>0.000022</td>\n",
       "      <td>19441</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AATCATGACCGCCCAAAC</td>\n",
       "      <td>1483</td>\n",
       "      <td>0.000456</td>\n",
       "      <td>8</td>\n",
       "      <td>AATCATGACCGCCCAAAC</td>\n",
       "      <td>274</td>\n",
       "      <td>0.000073</td>\n",
       "      <td>2597</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CACCGCGAGAGACACCTG</td>\n",
       "      <td>1381</td>\n",
       "      <td>0.000424</td>\n",
       "      <td>9</td>\n",
       "      <td>CACCGCGAGAGACACCTG</td>\n",
       "      <td>253</td>\n",
       "      <td>0.000068</td>\n",
       "      <td>3839</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GCCCCTGTTCGCCCAGGA</td>\n",
       "      <td>1372</td>\n",
       "      <td>0.000422</td>\n",
       "      <td>10</td>\n",
       "      <td>GCCCCTGTTCGCCCAGGA</td>\n",
       "      <td>562</td>\n",
       "      <td>0.000150</td>\n",
       "      <td>71</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12041</th>\n",
       "      <td>ACGGTTAAGACCGAAAAC</td>\n",
       "      <td>5</td>\n",
       "      <td>0.000002</td>\n",
       "      <td>61119</td>\n",
       "      <td>ACGGTTAAGACCGAAAAC</td>\n",
       "      <td>36</td>\n",
       "      <td>0.000010</td>\n",
       "      <td>20657</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12042</th>\n",
       "      <td>ATATGAGCAAAAAAAAGT</td>\n",
       "      <td>5</td>\n",
       "      <td>0.000002</td>\n",
       "      <td>61125</td>\n",
       "      <td>ATATGAGCAAAAAAAAGT</td>\n",
       "      <td>110</td>\n",
       "      <td>0.000029</td>\n",
       "      <td>17871</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12043</th>\n",
       "      <td>GCAAAACAGGGCAGTCAC</td>\n",
       "      <td>5</td>\n",
       "      <td>0.000002</td>\n",
       "      <td>61134</td>\n",
       "      <td>GCAAAACAGGGCAGTCAC</td>\n",
       "      <td>150</td>\n",
       "      <td>0.000040</td>\n",
       "      <td>11476</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12044</th>\n",
       "      <td>AAAAAAAGGATAAAAAAG</td>\n",
       "      <td>5</td>\n",
       "      <td>0.000002</td>\n",
       "      <td>61162</td>\n",
       "      <td>AAAAAAAGGATAAAAAAG</td>\n",
       "      <td>252</td>\n",
       "      <td>0.000067</td>\n",
       "      <td>3899</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12045</th>\n",
       "      <td>TGACTGAAAGGCCTAGTA</td>\n",
       "      <td>5</td>\n",
       "      <td>0.000002</td>\n",
       "      <td>61166</td>\n",
       "      <td>TGACTGAAAGGCCTAGTA</td>\n",
       "      <td>9</td>\n",
       "      <td>0.000002</td>\n",
       "      <td>21121</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>12046 rows Ã— 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              Selected BC  Selected Counts  Selected Freq  Selected Rank  \\\n",
       "0      AAAAGCGCCGACAAACCC             2271       0.000698              2   \n",
       "1      GAGGACGCCACAAACAGA             1524       0.000468              7   \n",
       "2      AATCATGACCGCCCAAAC             1483       0.000456              8   \n",
       "3      CACCGCGAGAGACACCTG             1381       0.000424              9   \n",
       "4      GCCCCTGTTCGCCCAGGA             1372       0.000422             10   \n",
       "...                   ...              ...            ...            ...   \n",
       "12041  ACGGTTAAGACCGAAAAC                5       0.000002          61119   \n",
       "12042  ATATGAGCAAAAAAAAGT                5       0.000002          61125   \n",
       "12043  GCAAAACAGGGCAGTCAC                5       0.000002          61134   \n",
       "12044  AAAAAAAGGATAAAAAAG                5       0.000002          61162   \n",
       "12045  TGACTGAAAGGCCTAGTA                5       0.000002          61166   \n",
       "\n",
       "               Control BC  Control Counts  Control Freq  Control Rank  \n",
       "0      AAAAGCGCCGACAAACCC             173      0.000046          8329  \n",
       "1      GAGGACGCCACAAACAGA              83      0.000022         19441  \n",
       "2      AATCATGACCGCCCAAAC             274      0.000073          2597  \n",
       "3      CACCGCGAGAGACACCTG             253      0.000068          3839  \n",
       "4      GCCCCTGTTCGCCCAGGA             562      0.000150            71  \n",
       "...                   ...             ...           ...           ...  \n",
       "12041  ACGGTTAAGACCGAAAAC              36      0.000010         20657  \n",
       "12042  ATATGAGCAAAAAAAAGT             110      0.000029         17871  \n",
       "12043  GCAAAACAGGGCAGTCAC             150      0.000040         11476  \n",
       "12044  AAAAAAAGGATAAAAAAG             252      0.000067          3899  \n",
       "12045  TGACTGAAAGGCCTAGTA               9      0.000002         21121  \n",
       "\n",
       "[12046 rows x 8 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Comparison of selected to control barcodes (Hamming distance of 1)\n",
    "# Returns dataframe of: shared BCs, read counts, frequencies, and BC ranks in both selected and control datasets\n",
    "rank_cutoff = 1000\n",
    "hamming_cutoff = 1\n",
    "unselected_1 = 'D21-9816'\n",
    "selected_1 = 'D21-9817'\n",
    "unstim_1 = 'D21-9818'\n",
    "unselected_2 = 'D21-9819'\n",
    "selected_2 = 'D21-9820'\n",
    "unstim_2 = 'D21-9821'\n",
    "\n",
    "selected_v_unstim_2 = barcode_overlap(selected_2, unstim_2, hamming_cutoff, read_cutoff = 5)\n",
    "selected_v_unstim_2\n",
    "save_output('Rep2_selected_unstim_overlap', selected_v_unstim_2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'./output/Rep1_selected_Rep2_selected_overlap.xlsx'"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "selected_2_v_1 = barcode_overlap(selected_2, selected_1, hamming_cutoff, read_cutoff = 5)\n",
    "selected_2_v_1\n",
    "save_output('Rep1_selected_Rep2_selected_overlap', selected_2_v_1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'./output/Rep2_unstim_v_stim.xlsx'"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "unstim2_v_stim2 = barcode_overlap(unstim_2, selected_2, hamming_cutoff, read_cutoff = 5)\n",
    "unstim2_v_stim2\n",
    "save_output('Rep2_unstim_v_stim', unstim2_v_stim2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'./output/Rep2_unstim_v_Rep1_stim.xlsx'"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "unstim2_v_stim1 = barcode_overlap(unstim_2, selected_1, hamming_cutoff, read_cutoff = 5)\n",
    "unstim2_v_stim1\n",
    "save_output('Rep2_unstim_v_Rep1_stim', unstim2_v_stim1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "triple_overlap = selected_2_v_1.merge(selected_v_unstim_2, on = 'Selected BC')\n",
    "triple_overlap"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "triple_common_bcs = list(triple_overlap['Selected BC'])\n",
    "shared_selection_bcs = selected_2_v_1\n",
    "for common_bc in triple_common_bcs:\n",
    "    index = shared_selection_bcs[shared_selection_bcs['Selected BC'] == common_bc].index.values\n",
    "    shared_selection_bcs = shared_selection_bcs.drop(index,'index')\n",
    "shared_selection_bcs\n",
    "save_output('selection_1_and_2_overlap_without_unstim', shared_selection_bcs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
